compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs

```diff
@@ -184,3 +199,170 @@ prettyFileDiagnostics (filePath, diags) =
+    dontKnow = "<unknown>"
+
+    getDiags :: DiagnosticsBySource -> [(T.Text, [LSP.Diagnostic])]
+    getDiags = map (\(ds, diag) -> (fromMaybe dontKnow ds, toList diag)) . Map.assocs
+
+emptyDiagnostics :: Diagnostics stage
+emptyDiagnostics = Diagnostics mempty
+
+-- | Sets the diagnostics for a file and compilation step
+--   if you want to clear the diagnostics call this with an empty list
+setStageDiagnostics ::
+  Show stage =>
+  FilePath ->
+  Maybe UTCTime ->
+  -- ^ the time that the file these diagnostics originate from was last edited
+  stage ->
+  [LSP.Diagnostic] ->
+  Diagnostics stage ->
+  IO (Diagnostics stage)
+setStageDiagnostics fp timeM stage diags (Diagnostics ds) = do
+    uri <- toUri fp
+    let thisFile = Map.lookup uri ds
+        addedStage :: StoreItem
+        addedStage = StoreItem (Just posixTime) $ storeItem thisFile
+    pure $ Diagnostics $ Map.insert uri addedStage ds where
+
+    storeItem :: Maybe StoreItem -> DiagnosticsBySource
+    storeItem = \case
+        Nothing ->
+            Map.singleton k v
+
+        Just (StoreItem Nothing _) ->
+            noVersions
+
+        Just (StoreItem (Just prevTime) bySource) ->
+            case compare posixTime prevTime of
+                EQ -> insert bySource
+                GT -> insert $ Map.map (SL.map tagOutdated)  bySource
+                LT -> bySource
+
+    -- Change the color of outdated diagnostics
+    tagOutdated :: Diagnostic -> Diagnostic
+    tagOutdated d = d{_severity = Just DsInfo}
+
+    isOutdated :: Diagnostic -> Bool
+    isOutdated Diagnostic{_severity} = _severity == Just DsInfo
+
+    removeOutdated, insert :: DiagnosticsBySource -> DiagnosticsBySource
+    removeOutdated = Map.filter (not . any isOutdated)
+
+    insert bySource =
+        let removeExpired =
+                if Map.lookup k bySource == Just (SL.map tagOutdated v)
+                then removeOutdated
+                else id
+        in Map.insert k v $ removeExpired bySource
+
+    k = Just $ T.pack $ show stage
+    v = SL.toSortedList diags
+
+    posixTime :: Int
+    posixTime = maybe 0 (fromEnum . utcTimeToPOSIXSeconds) timeM
+
+noVersions :: HasCallStack => a
+noVersions =
+    error "All store items must have versions"
+
+fromUri :: LSP.Uri -> FilePath
+fromUri = fromMaybe noFilePath . uriToFilePath
+
+-- TODO (MK) We canonicalize to make sure that the two files agree on use of
+-- / and \ and other shenanigans.
+-- haskell-lsp also has serious problems with relative file paths
+toUri :: FilePath -> IO LSP.Uri
+toUri = fmap filePathToUri . canonicalizePath
+
+getAllDiagnostics ::
+    Diagnostics stage ->
+    [FileDiagnostic]
+getAllDiagnostics =
+    concatMap (\(k,v) -> map (fromUri k,) $ getDiagnosticsFromStore v) . Map.toList . getStore
+
+getFileDiagnostics ::
+    FilePath ->
+    Diagnostics stage ->
+    IO [LSP.Diagnostic]
+getFileDiagnostics fp ds = do
+    uri <- toUri fp
+    pure $
+        maybe [] getDiagnosticsFromStore $
```

Comment:
`findWithDefault [] uri $ getStore ds` works nicely here.

Reply:
I don't think it does looking up by URI returns a `StoreItem` which needs to be unpacked by getDiagnostics from store, I could do this with an empty StoreItem as the default, but I don't think it's clearer

Reply:
Oh yeah, sorry missed that.